### ISSUE TITLE
Training exits after epoch 0 

### ISSUE THREAD

**Issue by gchal** [ucsdarclab/dVRL#2]
We really appreciate the effort you put in creating the "gym" interface of the dVRkit. However, we cannot run the simple ddpg+her algorithm, with your simulated environment. Since you have built on top of the Reacher environment,  running ddpg+her from openai baselines should have been trivial.
Apparently there is an issue, probably with the VREP communication when completing the environment step?
My  terminal's  output is: 
T: 100
_Q_lr: 0.001
_action_l2: 1.0
_batch_size: 256
_buffer_size: 1000000
_clip_obs: 200.0
_hidden: 256
_layers: 3
_max_u: 1.0
_network_class: baselines.her.actor_critic:ActorCritic
_norm_clip: 5
_norm_eps: 0.01
_pi_lr: 0.001
_polyak: 0.95
_relative_goals: False
_scope: ddpg
aux_loss_weight: 0.0078
bc_loss: 0
ddpg_params: {'buffer_size': 1000000, 'hidden': 256, 'layers': 3, 'network_class': 'baselines.her.actor_critic:ActorCritic', 'polyak': 0.95, 'batch_size': 256, 'Q_lr': 0.001, 'pi_lr': 0.001, 'norm_eps': 0.01, 'norm_clip': 5, 'max_u': 1.0, 'action_l2': 1.0, 'clip_obs': 200.0, 'scope': 'ddpg', 'relative_goals': False}
demo_batch_size: 128
env_name: dVRLReach-v0
gamma: 0.99
make_env: <function prepare_params.<locals>.make_env at 0x7efb47ca4400>
n_batches: 40
n_cycles: 50
n_test_rollouts: 10
noise_eps: 0.2
num_demo: 100
prm_loss_weight: 0.001
q_filter: 0
random_eps: 0.3
replay_k: 4
replay_strategy: future
rollout_batch_size: 1
test_with_polyak: False

*** Warning ***
You are running HER with just a single MPI worker. This will work, but the experiments that we report in Plappert et al. (2018, https://arxiv.org/abs/1802.09464) were obtained with --num_cpu 19. This makes a significant difference and if you are looking to reproduce those results, be aware of this. Please also refer to https://github.com/openai/baselines/issues/314 for further details.
****************

WARNING:tensorflow:From /baselines/baselines/her/ddpg.py:87: The name tf.variable_scope is deprecated. Please use tf.compat.v1.variable_scope instead.

Creating a DDPG agent with action space 3 x 1.0...
WARNING:tensorflow:From /baselines/baselines/her/util.py:66: dense (from tensorflow.python.layers.core) is deprecated and will be removed in a future version.
Instructions for updating:
Use keras.layers.dense instead.
2020-03-04 17:52:09.764217: W tensorflow/compiler/jit/mark_for_compilation_pass.cc:1412] (One-time warning): Not using XLA:CPU for cluster because envvar TF_XLA_FLAGS=--tf_xla_cpu_global_jit was not set.  If you want XLA:CPU, either set that envvar, or use experimental_jit_scope to enable XLA:CPU.  To confirm that XLA is active, pass --vmodule=xla_compilation_cache=1 (as a proper command-line flag, not via TF_XLA_FLAGS) or set the envvar XLA_FLAGS=--xla_hlo_profile.
Training...
---------------------------------
| epoch              | 0        |
| stats_g/mean       | -0.0325  |
| stats_g/std        | 0.279    |
| stats_o/mean       | 0.00811  |
| stats_o/std        | 0.0657   |
| test/episode       | 10       |
| test/mean_Q        | -3.08    |
| test/success_rate  | 0        |
| train/episode      | 50       |
| train/success_rate | 0        |
---------------------------------

Process finished with exit code 0

**Comment by SarahChiu** [comment]
From your terminal output, there seems to be no error related to the environment. 
How many time steps do you run for DDPG+HER?

**Comment by gchal** [comment]
> From your terminal output, there seems to be no error related to the environment.
> How many time steps do you run for DDPG+HER?
Thanks for the immediate reply. I run the command: 

python3 -m baselines.run --alg=her --env=dVRLReach-v0 --num_timesteps=5000 

which is the exact command for the respective FetchReach-v1 command in openai, which trains up to 50 epochs (i.e. 100 steps per epoch).

**Comment by SarahChiu** [comment]
If you take a look at her.py from openai baselines (https://github.com/openai/baselines/blob/master/baselines/her/her.py), the way they set their number of epochs is 
n_epochs = total_timesteps // n_cycles // rollout_worker.T // rollout_worker.rollout_batch_size. 

In your case, n_epochs will be 5000 / (50 * 100 * 1) = 1. That's why your program ends after running 1 epoch. 
You can try to increase the number of training time steps.

**Comment by gchal** [comment]
Thanks for the reply! I haven't noticed it that. strange because with the exact same command the FetchReacher trains for 50 epochs. Thanks for pointing that out. I would recommend that you include in your readme file that for applying the ddpg+her for openai someone should run for example: 
python3 -m baselines.run --alg=her --env=dVRLReach-v0 --num_timesteps=500000

**Comment by leey127** [comment]
Hello! I tried to run DDPG+HER algorithm from openai baselines, just like what you described. But I still got "train/success_rate=0" after so many epochs (By the way, the training speed was very low...). Have you got some results running this? Do I have to modify the parameters of the algorithm or something else?

Running `python -m baselines.run --alg=her --env=dVRLReach-v0 --num_timesteps=500000 --save_path=~/policies/her/dvrlreach5k --play`, the output after 41 epochs:
---------------------------------
| epoch              | 41       |
| stats_g/mean       | -0.167   |
| stats_g/std        | 0.653    |
| stats_o/mean       | 0.0307   |
| stats_o/std        | 0.192    |
| test/episode       | 420      |
| test/mean_Q        | -63.8    |
| test/success_rate  | 0        |
| train/episode      | 2.1e+03  |
| train/success_rate | 0        |
---------------------------------

**Comment by bango123** [comment]
Hello, typically I found each step to take 4-5ms (which is pretty slow). To speed up training, I recommend modifying the DDPG+HER baseline code such that you can collect multiple episodes worth of data in parallel by training on multiple environments.

With regards to convergence for DDPG+HER, I did not have to modify any of the original baselines parameters for reach. I recommend confirming your replay buffer is storing the state, action, next state, and reward tuple correctly. Then check if the hindsight experiences being generated are also correct. 

To convince yourself it actually can converge, a simple p-controller will work for reach.

**Comment by neriniccolodei** [comment]
> If you take a look at her.py from openai baselines (https://github.com/openai/baselines/blob/master/baselines/her/her.py), the way they set their number of epochs is
> n_epochs = total_timesteps // n_cycles // rollout_worker.T // rollout_worker.rollout_batch_size.
> 
> In your case, n_epochs will be 5000 / (50 * 100 * 1) = 1. That's why your program ends after running 1 epoch.
> You can try to increase the number of training time steps.

Dear @SarahChiu, do you perhaps have an explanation of these parameters n_cycles, rollout_worker.T and rollout_worker.rollout_batch_size? 
I would like to reinforce (sorry for the pun) my understanding with a source that explains their meaning. 

Best regards,
Neri

**Comment by SarahChiu** [comment]
Hi @nndei 
n_cycles is how many times the agent collects one episode in each training epoch. Its default value is 50. 
n_batches is how many times the networks are updated after the agent collects one episode. Its default value is 40. 
rollout_worker.T is the total time steps for each episode. It is defined in the gym environment. 

rollout_worker.rollout_batch_size will be the same as the number of environments running in parallel. 
In the current version of openai baselines, they provide a way to run multiple environments in parallel. You can set --num_env to be larger than 1. 
For example, if you set --num_env=5 and rollout_worker.T = 100, each time the agents generate rollouts, there will be a batch with size 100 * 5 being stored to the replay buffer.

**Comment by neriniccolodei** [comment]
Thank you, @SarahChiu.

May I ask you what is the difference between setting number of environments = 5 and for example running a training with the mpirun command, specifying 5 processes?

**Comment by SarahChiu** [comment]
Hi @nndei 
I updated my response in the previous post since there are some mistakes. Sorry about that. 

I haven't run baselines with multiple MPI workers so I am not quite sure about that. 
Yet I think [this post](https://github.com/openai/baselines/issues/485#issuecomment-413722708) might help.

**Comment by neriniccolodei** [comment]
> Hi @nndei
> I updated my response in the previous post since there are some mistakes. Sorry about that.
> 
> I haven't run baselines with multiple MPI workers so I am not quite sure about that.
> Yet I think [this post](https://github.com/openai/baselines/issues/485#issuecomment-413722708) might help.

Thank you, @SarahChiu.

If I may ask another thing, to my understanding the `train/success_rate` is not as important as the `test/success_rate` and shows the success rate of the task during the noisy training process. Therefore, it is likely less than 100% because there's noise and random actions to improve exploration.

When caring about plotting results, should I care about just the test success rate? Is it what the actual learned policy is capable of obtaining? 
Otherwise, I cannot understand why the plots in the dVRL paper show the pick reaching 100% success rate, if `train/success_rate` is likely not ever going to be 100%. 

I am basing such idea on this post: [https://github.com/openai/baselines/issues/468]

Best regards

**Comment by SarahChiu** [comment]
Hi @nndei, 

I think it is useful to observe both the training and testing success rate while training. 
Since one usually does not run many testing episodes during training (it can take long), the testing success rate might not be so accurate if the number of testing is not enough. 
Although random exploration affects the training success rate, the code in OpenAI baselines only allows 30% of random exploration. 
Thus, if the policy improves, the success rate should also gradually increase. 

Sometimes it is helpful to gradually decrease the random exploration effect while training, since if the policy learns something, there is no need for large exploration. 
For example, you can set the probability of providing random actions to 30% at the beginning and decrease it to 0% at the end of training. 
I believe this is what have been done in the dVRL paper, and that's why their success rate can reach 100%.

**Comment by neriniccolodei** [comment]
Thank you, @SarahChiu.

So, what is employed is straight up epsilon greedy, isn’t it? But there’s also noise added to the action as a Gaussian?

**Comment by SarahChiu** [comment]
Hi @nndei 

If epsilon represents the probability of no exploration, then its value increases. 
Noise is added to actions when there is exploration.

**Comment by neriniccolodei** [comment]
Thank you, so noise is added only when a random action is taken? Like a perturbation?

**Comment by SarahChiu** [comment]
Yes, exactly

**Comment by neriniccolodei** [comment]
Thank you for the help! I appreciate it
